cosmwasm/enclaves/shared/contract-engine/src/contract_validation.rs

```diff
@@ -675,24 +1151,167 @@ fn verify_message_params(
-            signer_addr, msg
-        );
-        return false;
+    info!("Verifying message sender...");
+    match msg {
+        CosmosSdkMsg::MsgRecvPacket { .. }
+        | CosmosSdkMsg::MsgAcknowledgement { .. }
+        | CosmosSdkMsg::MsgTimeout { .. } => {
+            // No sender to verify.
+            // Going to pass null sender to the contract if all other checks pass.
+        }
+        CosmosSdkMsg::MsgExecuteContract { .. }
+        | CosmosSdkMsg::MsgInstantiateContract { .. }
+        | CosmosSdkMsg::Other => {
+            if msg.sender() != Some(sender) {
+                warn!(
+                    "message sender did not match cosmwasm message sender: {:?} {:?}",
+                    sender, msg
+                );
+                return false;
+            }
+        }
     }
 
-    if !verify_contract(msg, contract_address) {
+    info!("Verifying contract address...");
+    if !verify_contract_address(msg, contract_address) {
         warn!("Contract address verification failed!");
         return false;
     }
 
-    info!("Verifying funds..");
-    if !verify_funds(msg, sent_funds) {
+    info!("Verifying sent funds...");
+    if !verify_sent_funds(msg, sent_funds) {
         warn!("Funds verification failed!");
         return false;
     }
 
     true
 }
+
+/// ReceiverChainIsSource returns true if the denomination originally came
+/// from the receiving chain and false otherwise.
+fn receiver_chain_is_source(source_port: &str, source_channel: &str, denom: &str) -> bool {
+    // The prefix passed in should contain the SourcePort and SourceChannel.
+    // If the receiver chain originally sent the token to the sender chain
+    // the denom will have the sender's SourcePort and SourceChannel as the
+    // prefix.
+    let voucher_prefix = get_denom_prefix(source_port, source_channel);
+    denom.starts_with(&voucher_prefix)
+}
+
+/// GetDenomPrefix returns the receiving denomination prefix
+fn get_denom_prefix(port_id: &str, channel_id: &str) -> String {
+    format!("{}/{}/", port_id, channel_id)
+}
+
+/// DenomTrace contains the base denomination for ICS20 fungible tokens and the
+/// source tracing information path.
+struct DenomTrace {
+    /// path defines the chain of port/channel identifiers used for tracing the
+    /// source of the fungible token.
+    path: String,
+    /// base denomination of the relayed fungible token.
+    base_denom: String,
+}
+
+impl DenomTrace {
+    /// Hash returns the hex bytes of the SHA256 hash of the DenomTrace fields using the following formula:
+    /// hash = sha256(tracePath + "/" + baseDenom)
+    pub fn hash(&self) -> Vec<u8> {
+        let hash = Sha256::digest(self.get_full_denom_path().as_bytes());
+        hash.to_vec()
+    }
+
+    /// IBCDenom a coin denomination for an ICS20 fungible token in the format
+    /// 'ibc/{hash(tracePath + baseDenom)}'. If the trace is empty, it will return the base denomination.
```

Comment:
Not sure what these types are doing here. Move them to a different file

Reply:
https://github.com/scrtlabs/SecretNetwork/commit/d487b138c